### PR TITLE
Correct template-specific workflow syntax

### LINKED ISSUE #5103: Correct use of $default-branch syntax for an example of a standard (non-template) workflow
### What article on docs.github.com is affected?

https://docs.github.com/en/actions/guides/building-and-testing-nodejs#starting-with-the-nodejs-workflow-template

### What part(s) of the article would you like to see updated?

The yaml example given for copy-paste in the [Starting with the Node.js workflow template](https://docs.github.com/en/actions/guides/building-and-testing-nodejs#starting-with-the-nodejs-workflow-template) section uses `$default-branch` as the branch filter within both the `push` and `pull-request` events.

The `$default-branch` syntax is _correct_ for a starter template like [the Node.js starter](https://github.com/actions/starter-workflows/blob/d9236ebe5585b1efd5732a29ea126807279ccd56/ci/node.js.yml) and is automatically changed to the repository's default branch on instantiating the template.

__However__, copy-pasting this syntax as suggested in this section creates an incorrect workflow.

In place of `$default-branch`, the code for copy-paste should use `main` (the current [default for the default branch](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-branches#about-the-default-branch)) and perhaps indicate that users need to edit this if they have a different default branch name.

### Additional information

Here is a commit showing that `main` syntax works but `$default-branch` syntax does not: https://github.com/steven-wolfman/test-workflow-template-syntax/commit/0d006c52a904911ddb71a13994a4864d7d022e1c#diff-fbdccf80e37a6640cb78bdcf9d90597cb91421747fd5c2365d878287e744ff99. (Click the check-mark to see that only the [workflow using standard syntax](https://github.com/steven-wolfman/test-workflow-template-syntax/blob/0d006c52a904911ddb71a13994a4864d7d022e1c/.github/workflows/standard.yml) runs while, e.g., the [workflow copy-pasted from the docs](https://github.com/steven-wolfman/test-workflow-template-syntax/blob/0d006c52a904911ddb71a13994a4864d7d022e1c/.github/workflows/copy-paste-of-nodejs-template.yml) does not run.

You can read a [discussion thread indicating that `$default-branch` is template-specific syntax](https://github.com/github/docs/discussions/5071).

Aside: If that raw YAML is somehow automatically-generated from the [Node.js starter](https://github.com/actions/starter-workflows/blob/d9236ebe5585b1efd5732a29ea126807279ccd56/ci/node.js.yml), which seems like a great idea, then this will require more care.

P.S. I'm putting this issue in as I prep a PR to correct this, but the contributing guidelines suggested starting with an issue. 😊

### PR DESCRIPTION
**Pull request by steven-wolfman**
### Why:

**Closes https://github.com/github/docs/issues/5103**

### What's being changed:

[The syntax `$default-branch` is specific to workflow templates only and cannot be used in standard workflows.](https://github.com/github/docs/discussions/5071) As a result, a copy-paste of the existing yaml workflow for "Starting with the Node.js workflow template" will fail. See https://github.com/steven-wolfman/test-workflow-template-syntax/commit/0d006c52a904911ddb71a13994a4864d7d022e1c#diff-fbdccf80e37a6640cb78bdcf9d90597cb91421747fd5c2365d878287e744ff99 for an example commit that triggers a workflow using the `main` syntax but fails to trigger workflows using the `$default-branch` syntax.

### Check off the following:
- [x] I have reviewed my changes in staging. (look for the **deploy-to-heroku** link in your pull request, then click **View deployment**)
- [x] For content changes, I have reviewed the [localization checklist](https://github.com/github/docs/blob/main/contributing/localization-checklist.md)
- [x] For content changes, I have reviewed the [Content style guide for GitHub Docs](https://github.com/github/docs/blob/main/contributing/content-style-guide.md).

### CODE REVIEW

File: content/actions/guides/building-and-testing-nodejs.md [hunk 1]
```diff
@@ -37,17 +37,17 @@ We recommend that you have a basic understanding of Node.js, YAML, workflow conf
 
 {% data variables.product.prodname_dotcom %} provides a Node.js workflow template that will work for most Node.js projects. This guide includes npm and Yarn examples that you can use to customize the template. For more information, see the [Node.js workflow template](https://github.com/actions/starter-workflows/blob/main/ci/node.js.yml).
 
-To get started quickly, add the template to the `.github/workflows` directory of your repository.
+To get started quickly, add the template to the `.github/workflows` directory of your repository. (If your default branch is not named `main`, change `main` to your default branch's name.)
```
Comment (steven-wolfman) [inline content/actions/guides/building-and-testing-nodejs.md:40]: I'd love feedback from a writer on how and whether to comment on what to do for people whose default branch is not named `main`.

**Comment by steven-wolfman** [comment]
OK. There are a few more places where this appears. In English:
+ https://github.com/steven-wolfman/docs/content/actions/guides/building-and-testing-ruby.md

Plus these translations of both the Ruby and Node.js documents:
+ https://github.com/steven-wolfman/docs/translations/ko-KR/content/actions/guides/building-and-testing-nodejs.md
+ https://github.com/steven-wolfman/docs/translations/ko-KR/content/actions/guides/building-and-testing-ruby.md
+ https://github.com/steven-wolfman/docs/translations/de-DE/content/actions/guides/building-and-testing-nodejs.md
+ https://github.com/steven-wolfman/docs/translations/de-DE/content/actions/guides/building-and-testing-ruby.md
+ https://github.com/steven-wolfman/docs/translations/pt-BR/content/actions/guides/building-and-testing-nodejs.md
+ https://github.com/steven-wolfman/docs/translations/pt-BR/content/actions/guides/building-and-testing-ruby.md
+ https://github.com/steven-wolfman/docs/translations/ru-RU/content/actions/guides/building-and-testing-nodejs.md
+ https://github.com/steven-wolfman/docs/translations/ru-RU/content/actions/guides/building-and-testing-ruby.md
+ https://github.com/steven-wolfman/docs/translations/es-ES/content/actions/guides/building-and-testing-nodejs.md
+ https://github.com/steven-wolfman/docs/translations/es-ES/content/actions/guides/building-and-testing-ruby.md
+ https://github.com/steven-wolfman/docs/translations/zh-CN/content/actions/guides/building-and-testing-nodejs.md
+ https://github.com/steven-wolfman/docs/translations/zh-CN/content/actions/guides/building-and-testing-ruby.md
+ https://github.com/steven-wolfman/docs/translations/ja-JP/content/actions/guides/building-and-testing-nodejs.md
+ https://github.com/steven-wolfman/docs/translations/ja-JP/content/actions/guides/building-and-testing-ruby.md

(A grep of all files for `$default-branch` shows only these locations plus _appropriate_ reference at https://github.com/steven-wolfman/docs/blob/main/content/actions/learn-github-actions/sharing-workflows-with-your-organization.md and its translations.)

I'll work on an update for https://github.com/steven-wolfman/docs/content/actions/guides/building-and-testing-ruby.md. I don't think I can reasonably help with the text in any of the translations.

File: content/actions/guides/building-and-testing-nodejs.md (hunk 1 above)
Comment (steven-wolfman) [reply content/actions/guides/building-and-testing-nodejs.md:40]: Whatever text ends up being used, the same text should appear in the Ruby version also changed in this commit.

**Review by Nirzak — COMMENTED** [review]
I think it's not compulsory to rename your default branch main. I have set "source' as my default branch and I have written the workflow as `on:
push:
branches: [source]` and the code just works fine for me. So, default-branch option is the correct one. It shouldn't be stated as 'main' as people can rename their default branch to any name.

**Comment by steven-wolfman** [comment]
> I think it's not compulsory to rename your default branch main. I have set "source' as my default branch and I have written the workflow as `on: push: branches: [source]` and the code just works fine for me. So, default-branch option is the correct one. It shouldn't be stated as 'main' as people can rename their default branch to any name.

I agree that it is not compulsory to name the default branch main.

The trouble is that the existing code shows `$default-branch` which does not work no matter what your default branch is named. Instead, that syntax is specific to GitHub workflow _templates_ and does not work in standard GitHub workflows.

I'm unsure whether putting `main` into the raw code is the right way to go, but leaving it as `$default-branch` with no further commentary seems clearly wrong.

Does that make sense?

**Comment by Nirzak** [comment]
> > I think it's not compulsory to rename your default branch main. I have set "source' as my default branch and I have written the workflow as `on: push: branches: [source]` and the code just works fine for me. So, default-branch option is the correct one. It shouldn't be stated as 'main' as people can rename their default branch to any name.
> 
> I agree that it is not compulsory to name the default branch main.
> 
> The trouble is that the existing code shows `$default-branch` which does not work no matter what your default branch is named. Instead, that syntax is specific to GitHub workflow _templates_ and does not work in standard GitHub workflows.
> 
> I'm unsure whether putting `main` into the raw code is the right way to go, but leaving it as `$default-branch` with no further commentary seems clearly wrong.
> 
> Does that make sense?

Yeah I agree but for that an elaborated documentation is needed. If we just put main in the raw code then many one will think they have to set main in their workflow. So the people who have default-branch named as "master" will face error.

**Comment by steven-wolfman** [comment]
> > I think it's not compulsory to rename your default branch main. I have set "source' as my default branch and I have written the workflow as `on: push: branches: [source]` and the code just works fine for me. So, default-branch option is the correct one. It shouldn't be stated as 'main' as people can rename their default branch to any name.
> 
> I agree that it is not compulsory to name the default branch main.
> 
> The trouble is that the existing code shows `$default-branch` which does not work no matter what your default branch is named. Instead, that syntax is specific to GitHub workflow _templates_ and does not work in standard GitHub workflows.
> 
> I'm unsure whether putting `main` into the raw code is the right way to go, but leaving it as `$default-branch` with no further commentary seems clearly wrong.
> 
> Does that make sense?

Just to make this more concrete, please do open https://github.com/steven-wolfman/test-workflow-template-syntax/commit/0d006c52a904911ddb71a13994a4864d7d022e1c

There, you can see that a workflow that directly copies-and-pastes the `$default-branch` syntax will not run on a push to the default branch. (Click the green checkmark to see what did run, which was the only one of the three workflows that uses `main`. The two workflows using `$default-branch` didn't run because the push was not to a branch named literally `$default-branch`. You can see the copy-pasted workflow that uses `$default-branch` at https://github.com/steven-wolfman/test-workflow-template-syntax/blob/0d006c52a904911ddb71a13994a4864d7d022e1c/.github/workflows/copy-paste-of-nodejs-template.yml. All of this would work the same with `source` as the default branch, replacing `main` where it appears, i.e., only one of the three workflows would run.)

**Comment by steven-wolfman** [comment]
> > > I think it's not compulsory to rename your default branch main. I have set "source' as my default branch and I have written the workflow as `on: push: branches: [source]` and the code just works fine for me. So, default-branch option is the correct one. It shouldn't be stated as 'main' as people can rename their default branch to any name.
> > 
> > 
> > I agree that it is not compulsory to name the default branch main.
> > The trouble is that the existing code shows `$default-branch` which does not work no matter what your default branch is named. Instead, that syntax is specific to GitHub workflow _templates_ and does not work in standard GitHub workflows.
> > I'm unsure whether putting `main` into the raw code is the right way to go, but leaving it as `$default-branch` with no further commentary seems clearly wrong.
> > Does that make sense?
> 
> Yeah I agree but for that an elaborated documentation is needed. If we just put main in the raw code then many one will think they have to set main in their workflow. So the people who have default-branch named as "master" will face error.

Totally, but isn't that better than the current case where _everyone_ will have an error?

And.. is the correct fix for the more elaborate documentation to clarify the need for rename just before the code block (which is what I tried)?

Thanks!

**Comment by Nirzak** [comment]
> > > > I think it's not compulsory to rename your default branch main. I have set "source' as my default branch and I have written the workflow as `on: push: branches: [source]` and the code just works fine for me. So, default-branch option is the correct one. It shouldn't be stated as 'main' as people can rename their default branch to any name.
> > > 
> > > 
> > > I agree that it is not compulsory to name the default branch main.
> > > The trouble is that the existing code shows `$default-branch` which does not work no matter what your default branch is named. Instead, that syntax is specific to GitHub workflow _templates_ and does not work in standard GitHub workflows.
> > > I'm unsure whether putting `main` into the raw code is the right way to go, but leaving it as `$default-branch` with no further commentary seems clearly wrong.
> > > Does that make sense?
> > 
> > 
> > Yeah I agree but for that an elaborated documentation is needed. If we just put main in the raw code then many one will think they have to set main in their workflow. So the people who have default-branch named as "master" will face error.
> 
> Totally, but isn't that better than the current case where _everyone_ will have an error?
> 
> And.. is the correct fix for the more elaborate documentation to clarify the need for rename just before the code block (which is what I tried)?
> 
> 

> > > > I think it's not compulsory to rename your default branch main. I have set "source' as my default branch and I have written the workflow as `on: push: branches: [source]` and the code just works fine for me. So, default-branch option is the correct one. It shouldn't be stated as 'main' as people can rename their default branch to any name.
> > > 
> > > 
> > > I agree that it is not compulsory to name the default branch main.
> > > The trouble is that the existing code shows `$default-branch` which does not work no matter what your default branch is named. Instead, that syntax is specific to GitHub workflow _templates_ and does not work in standard GitHub workflows.
> > > I'm unsure whether putting `main` into the raw code is the right way to go, but leaving it as `$default-branch` with no further commentary seems clearly wrong.
> > > Does that make sense?
> > 
> > 
> > Yeah I agree but for that an elaborated documentation is needed. If we just put main in the raw code then many one will think they have to set main in their workflow. So the people who have default-branch named as "master" will face error.
> 
> Totally, but isn't that better than the current case where _everyone_ will have an error?
> 
> And.. is the correct fix for the more elaborate documentation to clarify the need for rename just before the code block (which is what I tried)?
> 
> Thanks!

Yeah default-branch must have to be rename as the branch name that you've set as your default-branch. Set it as main in the raw code will help many ones because we know a huge number of github users are using main as their default-branch. But it still there should be extra documentation for those who have custom default-branch.

**Comment by janiceilene** [comment]
Thanks so much for opening an issue and a PR @steven-wolfman! You don't need to worry about any of the translations, updated content in English is automatically sent to our translators once it's merged. I'll get this triaged for review ⚡

**Comment by steven-wolfman** [comment]
Thanks!

---------------------------------
Steven A. Wolfman  Ph.D. (he/him)
Professor of Teaching
Department of Computer Science
The University of British Columbia | Vancouver Campus | Musqueam Traditional Territory
201-2366 Main Mall | Vancouver BC | V6T 1Z4 Canada
***@***.***

**Review by lucascosti — APPROVED** [review]
Thanks for this, @steven-wolfman! I slightly reworded the mentioning of `main` before the workflow, and replaced a few other instances of `$default-branch` in the ruby article.

I'll get this merged in! 🎉

**Pull request merged**